### DISCUSSION TITLE
Quick installation by GitHub Actions already available / planned?

### DISCUSSION THREAD

**Discussion by kevinmatthes** [sourcefrog/cargo-mutants#138] [Q&A] [answered]
Hello and thank you for this software!

I would like to enhance the CI of one of my Rust projects by adding this tool to that project's CI and now I was wondering whether there might be already a GitHub Action existing to quickly install the binary.

So far, I worked with https://github.com/taiki-e/install-action which will install the specified Rust binary using caches such that a notable speed-up of the CI can be achieved.  The limitation of this Action is that the Rust binary to be installed first of all needs to be hard-coded into the Action's definition, so it is not possible to install an arbitrary tool from crates.io.  As far as I can see, it does not support `cargo-mutants`, yet, so I would like to ask whether someone already created an according GitHub Action for `cargo-mutants` or the integration with any existing Action might be already planned / in progress.

I am looking forward to your feedback!

**Comment by kevinmatthes** [comment, marked as answer]
I found the following Action which solves this question for me:  https://github.com/baptiste0928/cargo-install.

```yml
      - uses: baptiste0928/cargo-install@v2.1.0
        with:
          crate: cargo-mutants
```

**Comment by sourcefrog** [comment]
That's great. I was just thinking about the same thing. Maybe we could get a single action that installs and runs cargo-mutants, and potentially installs a binary to make it a bit faster...

**Reply by kevinmatthes** [reply]
As a starting point for that, I would like to suggest to analyse how the other Rust binaries distributed by https://github.com/taiki-e/install-action are creating their releases.  The projects I have seen so far attach pre-compiled binaries to their releases to be downloaded by the Action using GitHub's API.  That way, the software is available very quickly, usually also for all default hosted runners.

As a first step in order to prepare a similar approach for `cargo-mutants`, I would recommend to have a look at possibilities for automation regarding the GitHub release creation.  For my own projects, I usually apply https://github.com/softprops/action-gh-release which has plenty of configuration options, including the possibility to attach some pre-compiled binaries.

Once the prepared executables can be downloaded, it is rather simple to create an Action.  For the beginning, a composite Action would suffice to simply grab the respective binary for the runner (the OS should be a configuration option of the Action) and execute the default `cargo mutants` command; it is more secure to hard-code the executed command and assuming the users to have a configuration file provided.  For convenience, since the binaries are already existing, then, this project could also be registered in bundling Actions like the already mentioned https://github.com/taiki-e/install-action which has about 3.5 thousands users.

**Reply by sourcefrog** [reply]
I set up [cargo-dist](https://opensource.axo.dev/cargo-dist/) which seems to do a good job at attaching binaries to releases, and that seems to work well with install-action to pull those binaries into workflows that use cargo-mutants, as in <https://github.com/sourcefrog/conserve/pull/239>. With this fixed it takes only about a second to install the binary.

Some follow-on things:
- suggest install-action in the docs
- add cargo-mutants to the list of tools that install-action knows about
